tests/unit/fem/test_build_dof_to_arrays.cpp

```diff
@@ -212,7 +244,10 @@ TEST_CASE("Build Dof To Arrays (Parallel)",
             ParFiniteElementSpace fespace(&pmesh, fec);
             HYPRE_Int size = fespace.GetTrueVSize();
 
-            fespace.BuildDofToArrays();
+            // Get all boundary ldofs
+            Array<int> bdr(1); bdr = 1;
+            fespace.GetEssentialVDofs(bdr, all_bdr_ldofs_marked);
+            FiniteElementSpace::MarkerToList(all_bdr_ldofs_marked, all_bdr_ldofs);
```

Comment:
Replace this by `GetEssentialTrueDofs()`, which returns a list directly, but I am not sure if that is the problem really. The split by METIS should be element-wise, so boundary should be always own 🤔 .

Reply:
Only issue however is that I need the ldofs, not the tdofs. I am looking into this myself now as well -- I think that iit *might* be a negative `int bel` being sent to `GetBdrElementDofs`

Reply:
Yes, that is probably the reason, because `all_bdr_ldofs` includes -1 probably, so it might come from here. There is no difference between TDof and VDof, they are just "owned", while VDofs might be ghosts (i.e. neighbor's).

Reply:
Btw, it would be better to not use the name "LDofs" here as it might be confused with the local dofs in the sense below.

Reply:
LDof is normally used in the sense of the local index within an *element*, so it is below the index to the array of the element Dofs. Please rename it here to TDofs 😉 .

Reply:
@najlkin This confusion was actually part of the doc change I clarified in the `GetLocalDofForDof` and `GetBdrLocalDofForDof` earlier in this PR --> I'm using the ldof definition [here](https://docs.mfem.org/html/classmfem_1_1FiniteElementSpace.html#ldof), where "local" refers to local on a given MPI proc. I had been quite confused when interpreting `GetLocalDofForDof` until digging in and seeing it took as an argument an ldof in the context of local to a rank as defined in the link above.

I've always treated ldof indices and tdof indices as generally disjoint (in that a tdof index and an ldof index for a given non-ghost DOF may be entirely different) - is this not the case?

I did try implementing the changes you suggested using `GetEssentialTrueDofs` but am still getting similar errors.

Reply:
I do not know who wrote that new documentation, but it is completely misleading 🤯 . We should fix it, unless there is a shift in the terminology, but obviously `GetLocalDofForDof()` does not follow that. @tzanio , do you know where this terminology comes from?

Reply:
In any case, @j-signorelli , have you confirmed that you have `-1` in the list returned by `GetEssentialTrueDofs()`?

Reply:
Ohh, actually it can be, because you take just the first boundary, but it might be neighbors's if you are in a corner, i.e. there are more boundaries. You must overwrite the array if the associated boundary element is `-1` 😮 .

Reply:
Regarding the terminology, I just spoke with @tzanio and "local dof" should be primarily used in the sense of the MPI rank local. Unfortunately, the other use is also relatively common. It would be better to use a different name like element index or dof (but element dof is usually understood as a dof for an element), but it is how it is.

Reply:
Sorry for the late response, @najkin is correct in his description above ☝️ 

We have a better/more clear way to think about this, but that hasn't made it into the code.

Specifically, in the [operator decomposition](https://mfem.org/performance/#finite-element-operator-decomposition) where we have `T`-, `L`-, `E`-vectors we can talk about their entries as `TDofs`, `LDofs` and `EDofs` and each case we can refer to either global number of the dof or the local number inside the appropriate object -- MPI ranks for `TDofs` and `LDofs` and element for `EDofs`.

That being said, @j-signorelli -- do you understand what is causing the issue and how to fix it? Could it be that some of the MPI ranks are empty and that is not handled?

Reply:
> In any case, @j-signorelli , have you confirmed that you have `bdr_el==-1` for DOFs from the list returned by `GetEssentialTrueDofs()`?

Yes I can confirm that this is occurring when I use both `GetEssentialVDofs()` with `MarkerToList()` (LDofs) or `GetEssentialTrueDofs()` (tdofs). The `GetBdrElementForDof()` method is returning -1 for an index that is, according to those two methods, a bdr dof. When I add in the `continue` statement as you suggested below, the test still is not completing due to this error:

      Abort(134391310) on node 1 (rank 1 in comm 0): Fatal error in internal_Waitany: Message truncated, error stack:
      internal_Waitany(108)....: MPI_Waitany(count=2, array_of_requests=0x62edb1b1bd80, indx=0x7ffc72e5dacc, status=0x1) failed
      MPIR_Waitany(1281).......: 
      MPIDIG_recv_type_init(72): Message from rank 0 and tag 40822 truncated; 16 bytes received but buffer size is 36

This seems to be occurring during mesh decomposition. Notably I'm also getting these warnings:

    MFEM Warning: Non-positive attributes on the boundary!
     ... in function: virtual void mfem::Mesh::SetAttributes()
     ... in file: /home/j-signorelli/software/MFEM/git_repo/mesh/mesh.cpp:1619

but I am not really sure why because I believe I am properly setting all boundary elements to have an attribute of 1 at the end of `GetMesh()`:

    for (int i = 0; i < mesh->GetNBE(); i++) { mesh->SetBdrAttribute(i, 1); }


When I comment out the above line, I no longer get the fatal error. So just to summarize for completeness:

1. Including `continue` to prevent the seg fault yields another error I pasted above -- this new error doesn't occur if I comment out the `SetBdrAttribute` loop.
2. The non-positive attributes warning is showing when it really should not be as I am specifying them all to be 1.
3. The original error persists regardless and I am still unsure why - it may have something to do with the above problems.

Reply:
Only `MeshType::SEGMENT` is failing to properly set the bdr attributes. Instead of 1, it is being set to -1.

Reply:
Hmm, this is interesting, it seems there is a bug somewhere in `GenerateBoundaryElements()`, `SetBdrAttribute()` or `FinalizeTopology()` 🤔 .

Reply:
@najkin I am not sure why only segment was failing, but [61a49b8](https://github.com/mfem/mfem/pull/4416/commits/61a49b8ca080a4522fb5db431542043f7a04f69b) seems to have resolved this problem.

Reply:
But `FinalizeTopology()` has the default argument set to `true`, so I do not see how that can help 🤔 .

Reply:
Btw, please mind the spelling, it is `najlkin`, not `najkin`. That other guy got a message instead again 😄 .

Reply:
Ok, so what I have tested so far:

- `mesh->FinalizeTopology()` generates boundaries correctly, so I do not understand why there was a need to add anything as [you mentioned before](https://github.com/mfem/mfem/pull/4416#discussion_r1722117704) ?
- `mesh->GenerateBoundaryElements()` *before* `mesh->FinalizeTopology()` does something strange, I need to analyze that... 🤔 
- `mesh->GenerateBoundaryElements()` *after* `mesh->FinalizeTopology()` (`mesh->FinalizeTopology(false)`) works correctly though 😉

Reply:
@najlkin My apologies on the spelling! Oh I missed that regarding the default argument already being true... Then `GenerateBoundaryElements` must be problematic in some way with `MeshType::SEGMENT`.

Reply:
@najlkin Interesting, thank you for your assistance -- I hadn't realized that bdr elements would be automatically included.

I'm currently looking into why this MPI Waitany error is arising.

Reply:
In any case, are you getting any errors with the current version of the unit test? If not, then we should try to remerge it to `next`, because it generates massive error logs 😄 .

Reply:
I am still getting errors whenever I run on >= 3 mpi procs.

At this point, I have determined that the MPI error occurs ONLY after a `bdr_e==-1` situation occurs, where:

    if (bdr_e < 0 || bdr_e >= pmesh.GetNBE()) { num_bdr_elem_fails++; continue; }

evaluates to true on a given rank. When this happens, that MPI error occurs on that same rank for the *next* mesh decomposition here:

    ParMesh pmesh(MPI_COMM_WORLD, *mesh);

 I have no clue why this is occurring.

Reply:
I actually had thought that too but I got rid of the `continue` and instead changed it to:

    bdr_e = fespace.GetBdrElementForDof(all_bdr_ldofs[0]);
    bdr_l = fespace.GetBdrLocalDofForDof(all_bdr_ldofs[0]);
    
inside of the conditional and this same error was still occurring! I also didn't see any communications going on there either.

Reply:
About that `GenerateBoundaryElements()`, it should not be called before finalization of topology, because it actually just deletes the boundary structures as it cannot generate the boundaries correctly yet. Than it was actually `FinalizeTopology()` generating the boundary, but because of that deleted boundary structures, it copies the face elements from inside, which have -1 attribute for segment for some reason, so it becomes the boundary attribute. I do not know, maybe something more is going on, but it is a minor bug only.

Reply:
@termi-official , do you know why `Point` has the default attribute -1 unlike everything else (`Segement`, `Triangle`,...)? (@tzanio said you might know 😅 )

Reply:
From comparing the git blames I think it might just have been forgotten on the update to 1.2 https://github.com/mfem/mfem/commit/29ae8fb6ac9fce927fdfd01ef796f9df38f76e97

Reply:
Hmm, in other primitives than `Segment` it seems to be from version 1.0. Just that one was changed from -1 to 1 🤔 . But @tzanio does not know, so it is long buried in history why it was/is like that. I will open a separate PR for that, thanks 😉 .

Reply:
Btw, that is why there is: `mesh->AddBdrPoint(0);` , which is/was a workaround for the boundary attributes.